### PR TITLE
Add ability to specify service names by domain

### PR DESCRIPTION
**Pull request by stormsilver**
## What?
When using the `faraday` and `excon` contribs with the `split_by_domain` feature turned on, allow users to specify an additional hash that can override the service name by domain.

## Why?
We use an external service that makes a per-customer domain name. When we create HTTP requests to this service for our customers, it looks like `user-123.foo.com` and `user-abc.foo.com` (and so on). When we split by domain (which we want to do for all the _other_ HTTP services we use), it ends up creating hundreds of separate "services" on the service map - all of which are really the same external service. We'd like to have these all show up as one service.

## How?
By providing a hash in the form of `{/regex/ => 'desired_service_name'}`, we can match the desired domain names (`/user-[^.]+\.foo\.com/`) and provide the service name we want to see instead (`user.foo.com`)

### CODE REVIEW

**Comment by delner** [comment]
Thanks for the PR @stormsilver! I've been wanting to do something like this for a while, so I'm really excited to see this.

Couple of quick thoughts:

 - We probably would like to apply this behavior more generally to other HTTP integrations, too. Maybe in this PR if it ends up not being much work, but maybe in another PR if it is. It'd be cool if the mapping/configuration option were generic enough that we could apply it to other integrations easily.
 - We have a configuration multiplexing feature we might want to leverage to do this. The idea is you'd do something like `Datadog.configure[:faraday, /regex/][:service_name] = 'custom-service-name'` rather than having to define and provide a map option. We currently do this in ActiveRecord and in Dalli; see the [code that would enable this](https://github.com/DataDog/dd-trace-rb/blob/master/lib/ddtrace/contrib/configurable.rb#L34).

I think is definitely something we'd like to merge though those things considered.

**Comment by delner** [comment]
So I took a crack at implementing this via resolver pattern instead, and came up with this. You would use it like:

```ruby
Datadog.configure do |c|
  c.use :faraday, describe: /example\.com/ do |faraday|
    faraday.service_name = 'bar'
    faraday.split_by_domain = false # Only necessary if split_by_domain is true by default
  end
end
```

Tests pass so it seems to work. If we would want to incorporate this into the PR, some other things we'd have to do is:

 - Consider changing how the `RegexpResolver` has keys to it. Right now if you call it ever with an `Regexp`, it will cache that `Regexp` (e.g. `Datadog.configuration[:faraday, /some exp/]`). There might be a better way to be explicit about adding vs accessing that's safer.
 - Add unit tests for it.
 - Implement it for the other HTTP libraries.
 - Update the documentation.

I didn't have permissions to push to your forked repo, so I'm sharing the diff here:

[regexp_resolver.txt](https://github.com/DataDog/dd-trace-rb/files/3953370/regexp_resolver.txt)

Feedback welcome!

**Comment by stormsilver** [comment]
I applied you patch locally and spent some time spelunking the code to see how the resolvers work. This makes sense to me. 

Implementing this on all 4 of the HTTP libraries will require some solid refactoring of ethon and http. It makes the most sense to me to begin extracting stuff out into an HTTP tracer helper module. 

I will work on this in my spare time. It’ll be a bit before it’s done. 

-- Eric

**Comment by delner** [comment]
No worries, let me know if there's anything I can be of help with, particularly as it pertains to the resolver/configuration pattern.

File: lib/ddtrace/contrib/configurable.rb [hunk 1]
```diff
@@ -30,12 +30,13 @@ def configurations
           end
         end
 
-        def configure(key, options = {}, &block)
-          key = resolver.resolve(key || :default)
+        def configure(key = :default, options = {}, &block)
+          resolver.add_key(key)
```
Comment (stormsilver) [inline lib/ddtrace/contrib/configurable.rb]: This is how I ended up solving the awkward dance that `RegexpResolver` was doing to remember patterns.

File: lib/ddtrace/contrib/ethon/easy_patch.rb [hunk 2]
```diff
@@ -23,14 +27,13 @@ def http_request(url, action_name, options = {})
           end
 
           def headers=(headers)
-            return super unless tracer_enabled?
```
Comment (stormsilver) [inline lib/ddtrace/contrib/ethon/easy_patch.rb]: I had to remove the `tracer_enabled?` call here and in `reset` because, since we added the resolver pattern, you now have to have a hostname in order to determine the configuration (to decide if the tracer is enabled. theoretically you can now disable the tracer for requests matching `/do-not-want.io/`)... but you can set headers before supplying a URL to net/http.

This will have the side effect of creating a few DD instance variables on the net/http object even if the tracer is disabled.

File: lib/ddtrace/contrib/http/instrumentation.rb [hunk 3]
```diff
@@ -63,49 +72,57 @@ def request(req, body = nil, &block) # :yield: +response+
...
+            return unless analytics_enabled?(request_options)
+            Contrib::Analytics.set_sample_rate(span, analytics_sample_rate(request_options))
+          end
+
+          def datadog_pin(config = Datadog.configuration[:http])
             @datadog_pin ||= begin
-              service = Datadog.configuration[:http][:service_name]
-              tracer = Datadog.configuration[:http][:tracer]
+              service = config[:service_name]
+              tracer = config[:tracer]
```
Comment (stormsilver) [inline lib/ddtrace/contrib/http/instrumentation.rb]: There's a subtle bug here in that someone could call `datadog_pin` on this object with no args before running a request. It would then use the `:default` configuration to supply args to the pin. If the request is then run for a hostname which matches a regex configuration, we'll use the wrong configuration since it's been memoized.

I didn't have enough context to know if this is actually a problem, or what people use custom service names on pins for, or if the pin should just be removed since they look deprecated, or ...

**Comment by stormsilver** [comment]
Well, the CI tests failed which led me to realize that I had put the ethon tests in the wrong spot and they weren't even running except on CI (I thought I had run them) which led me to realize that _I hadn't even implemented the service_name change for ethon_, which is really embarrassing. So these failures are quite legit.

**Comment by delner** [comment]
Haha, that's quite already @stormsilver. I'll try to take a look at your changes see if I can give any helpful feedback.

File: lib/ddtrace/contrib/http/instrumentation.rb (hunk 3 above)
Comment (delner) [reply lib/ddtrace/contrib/http/instrumentation.rb]: Yeah I think this could be a problem, given that someone might build an HTTP object, set something on its `datadog_pin` then call it with a request resulting in this scenario.

I think that's one of the tricky things about this "by hostname" configuration: we have 1) default, 2) "by host" and 3) by object configuration which all has to be merged together to apply the settings that make the most sense for usability. Inevitably I think this resolution has to happen at request time (less than ideal, but unavoidable.)

File: lib/ddtrace/contrib/http/instrumentation.rb [hunk 4]
```diff
@@ -28,8 +29,12 @@ def self.after_request(&block)
...
+          include Datadog::Contrib::HttpAnnotationHelper
+
           def request(req, body = nil, &block) # :yield: +response+
-            pin = datadog_pin
+            host, = host_and_port(req)
+            request_options = datadog_configuration(host)
+            pin = datadog_pin(request_options)
             return super(req, body, &block) unless pin && pin.tracer
 
             if Datadog::Contrib::HTTP.should_skip_tracing?(req, @address, @port, pin.tracer)
```
Comment (delner) [inline lib/ddtrace/contrib/http/instrumentation.rb]: Something that's really cool about this resolver is we might be able to use it to remove this circuit breaker (which exists to prevent our tracer from tracing itself when it writes traces over HTTP to the agent.) Alternatively, we could probably use `enabled` as well to effect the same result.

Probably should be addressed in another PR though.

File: lib/ddtrace/contrib/configurable.rb [hunk 5]
```diff
@@ -30,12 +30,13 @@ def configurations
           end
         end
 
-        def configure(key, options = {}, &block)
-          key = resolver.resolve(key || :default)
+        def configure(key = :default, options = {}, &block)
+          resolver.add_key(key) unless key == :default
```
[the remark this answers is not in the document]
Comment (stormsilver) [reply lib/ddtrace/contrib/configurable.rb]: My feeling was that the codebase already treats `:default` as something special that shouldn't be remapped, so I didn't want to mess with that.

File: lib/ddtrace/contrib/configuration/resolvers/regexp_resolver.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+module Datadog
+  module Contrib
+    module Configuration
+      # Resolves a value to a configuration key
+      module Resolvers
+        # Matches strings against Regexps.
+        class RegexpResolver < Datadog::Contrib::Configuration::Resolver
+          def add_key(pattern)
+            patterns << pattern.is_a?(Regexp) ? pattern : /#{pattern}/
```
[the remark this answers is not in the document]
Comment (stormsilver) [reply lib/ddtrace/contrib/configuration/resolvers/regexp_resolver.rb]: Ah, yes. `Regexp.quote` is what you're talking about, I think: https://ruby-doc.org/core-2.7.0/Regexp.html#method-c-quote

File: lib/ddtrace/contrib/ethon/easy_patch.rb (hunk 2 above)
Comment (stormsilver) [reply lib/ddtrace/contrib/ethon/easy_patch.rb]: My thought was that it's better to respect all settings per host and leave a few extra ivars on the request object rather than have a surprising situation where you disabled the tracer for certain hosts but they are still getting traced (which would appear to the casual observer to be a bug).

I can certainly derive `enabled?` from the default config if there's no hostname if you feel like the tradeoff isn't worth it.

**Comment by stormsilver** [comment]
Okay @delner I think this is ready for another review.

**Comment by delner** [comment]
Awesome, I'll give it another look.

File: lib/ddtrace/contrib/configuration/resolvers/regexp_resolver.rb [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+module Datadog
+  module Contrib
+    module Configuration
+      # Resolves a value to a configuration key
+      module Resolvers
+        # Matches strings against Regexps.
+        class RegexpResolver < Datadog::Contrib::Configuration::Resolver
+          def add_key(pattern)
+            patterns << pattern.is_a?(Regexp) ? pattern : /#{Regexp.quote(pattern)}/
```
Comment (delner) [inline lib/ddtrace/contrib/configuration/resolvers/regexp_resolver.rb]: Although its kind of nice to convert everything to a Regex for consistency I think there's a couple of problems with interpolating strings to `Regexp`:

1. String values might match inadvertently, e.g. passing `'a'` would match `'a'` and `'aa'`.
2. There's an added expense to evaluating `Regexp` when doing `==` with a string might suffice.

I would recommend accepting any object, then in `#resolve` do a case statement such that if the pattern is a `Regexp`, do `=~`, otherwise coerce the pattern to a string and do `==`.

File: lib/ddtrace/contrib/configuration/resolvers/regexp_resolver.rb [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+          end
+
+          def resolve(name)
+            return :default if name == :default
+
+            name = name.to_s
+            matching_pattern = patterns.find { |p| p =~ name }
```
Comment (delner) [inline lib/ddtrace/contrib/configuration/resolvers/regexp_resolver.rb]: Maybe this isn't a big deal, but should we change `name` to `name.to_s`? Coercion might prevent some type incompatibility issues such as with `URI`, etc.

File: lib/ddtrace/contrib/http_annotation_helper.rb [hunk 9]
```diff
@@ -0,0 +1,10 @@
+module Datadog
+  module Contrib
+    # Contains methods helpful for tracing/annotating HTTP request libraries
+    module HttpAnnotationHelper
```
Comment (delner) [inline lib/ddtrace/contrib/http_annotation_helper.rb]: Should we extract this?

Although it does reduce some repetition, it might be a bit premature given its this single line function. I might favor the repetition over having the additional constant in this case; the use of options specific to the integration (such as `service_name` and `split_by_domain` I think reinforces this right now, too. Eventually if the behavior grows in scope, an extraction like this might be appropriate.

**Comment by delner** [comment]
To support a similar "multiplexing" scenario where we would benefit from the resolver changes, I've extracted those changes to #940 and merged them to `master`. We can rebase this PR on those changes which should reconcile things a bit.

**Comment by delner** [comment]
Hey @stormsilver , looks like this PR has grown stale, but we still wanted to see these changes incorporated. Given your repo doesn't grant push permissions to upstream repo members, I opened #953 which replaces this PR to rebase your commits and apply a few changes of my own.

We'll continue working on this feature in that PR; please feel free to review/contribute to that as you wish. Thanks again for your contributions so far!
